api-put-object-streaming.go

```diff
@@ -446,6 +449,193 @@ func (c *Client) putObjectMultipartStreamOptionalChecksum(ctx context.Context, b
 	return uploadInfo, nil
 }
 
+// putObjectMultipartStreamParallel uploads opts.NumThreads parts in parallel.
+// This is expected to take opts.PartSize * opts.NumThreads * (GOGC / 100) bytes of buffer.
+func (c *Client) putObjectMultipartStreamParallel(ctx context.Context, bucketName, objectName string,
+	reader io.Reader, opts PutObjectOptions) (info UploadInfo, err error) {
+	// Input validation.
+	if err = s3utils.CheckValidBucketName(bucketName); err != nil {
+		return UploadInfo{}, err
+	}
+
+	if err = s3utils.CheckValidObjectName(objectName); err != nil {
+		return UploadInfo{}, err
+	}
+
+	if !opts.SendContentMd5 {
+		if opts.UserMetadata == nil {
+			opts.UserMetadata = make(map[string]string, 1)
+		}
+		opts.UserMetadata["X-Amz-Checksum-Algorithm"] = "CRC32C"
+	}
+
+	// Cancel all when an error occurs.
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	// Calculate the optimal parts info for a given size.
+	partSize := int64(maxPartSize)
+	if opts.PartSize > 0 {
+		partSize = int64(opts.PartSize)
+	}
```

Comment:
we should be using this?

```
        // Calculate the optimal parts info for a given size.
        totalPartsCount, partSize, lastPartSize, err := OptimalPartInfo(size, opts.PartSize)
        if err != nil {
                return UploadInfo{}, err
        }
```

maxPartSize is 5Gi by default which is a bit too much?